### PR TITLE
remove new dictionaryexpression and keep as snippet only

### PR DESCRIPTION
**Pull request by m-nash**
Fixes https://github.com/microsoft/typespec/issues/3471

### CODE REVIEW

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/DictionaryInitializerExpression.cs [hunk 1]
```diff
@@ -5,29 +5,33 @@
 
 namespace Microsoft.Generator.CSharp.Expressions
 {
-    public sealed record DictionaryInitializerExpression(IReadOnlyList<(ValueExpression Key, ValueExpression Value)>? Values = null) : ValueExpression
+    public sealed record DictionaryInitializerExpression(IReadOnlyDictionary<ValueExpression, ValueExpression> Values) : ObjectInitializerExpression(Values)
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/DictionaryInitializerExpression.cs:8]: well changing it from a list of keyvalue pair to a dictionary might make the order of writing messy, the dictionary does not preserve the order of inputs.
Would this be a problem?

Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/DictionaryInitializerExpression.cs:8]: I don't know how the order of initializations would matter

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/NewInstanceExpression.cs [hunk 2]
```diff
@@ -5,20 +5,30 @@
...
-            if (InitExpression is { Parameters.Count: > 0 })
+            if (InitExpression is not null)
             {
+                if (!InitExpression.UseSingleLine)
+                {
+                    writer.WriteLine();
+                }
+                else
+                {
+                    writer.AppendRaw(" ");
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/NewInstanceExpression.cs:30]: nit: can we use `string.Empty` here?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/NewInstanceExpression.cs [hunk 3]
```diff
@@ -5,20 +5,30 @@
...
         private const int SingleLineParameterThreshold = 6;
 
         internal override void Write(CodeWriter writer)
         {
-            writer.Append($"new {Type}");
-            if (Parameters.Count > 0 || InitExpression is not { Parameters.Count: > 0 })
+            writer.Append($"new");
+            if (Type is not null)
+                writer.Append($" {Type}");
+            if (Parameters.Count > 0 || InitExpression is not { Values.Count: > 0 })
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/NewInstanceExpression.cs:17]: Should we use `Values.Any` instead?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/ObjectInitializerExpression.cs [hunk 4]
```diff
@@ -10,48 +10,45 @@ namespace Microsoft.Generator.CSharp.Expressions
     /// </summary>
     /// <param name="Parameters">The parameters to initialize the object to.</param>
     /// <param name="UseSingleLine">Flag to determine if the object should be initialized inline.</param>
-    public sealed record ObjectInitializerExpression(IReadOnlyDictionary<string, ValueExpression>? Parameters = null, bool UseSingleLine = true) : ValueExpression
```
Comment (jorgerangel-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/ObjectInitializerExpression.cs:13]: can you please update the method summary parameters?

**Review by jorgerangel-msft — APPROVED** [review]

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/NewInstanceExpression.cs (hunk 2 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/NewInstanceExpression.cs:30]: this isn't empty its a single space character.

Comment (jorgerangel-msft) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/NewInstanceExpression.cs:30]: ah you're right. Ignore this comment then.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/NewInstanceExpression.cs (hunk 3 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Expressions/NewInstanceExpression.cs:17]: typically only do that when we have an IEnumerable and don't know the implementation as to whether it has a `Count` or `Length`.  In this case we know they have a `Count` and can use that property directly.

**Pull request merged**